packages/lodestar/src/db/api/beacon/repositories/blockArchive.ts

```diff
@@ -34,20 +34,37 @@ export class BlockArchiveRepository extends BulkRepository<SignedBeaconBlock> {
     upperLimit: number | null,
     step: number | null = 1
   ): Promise<SignedBeaconBlock[]> {
+    const result = [];
+    for await (const signedBlock of this.getAllBetweenStream(lowerLimit, upperLimit, step)) {
+      result.push(signedBlock);
+    }
+    return result;
+  }
+
+  public getAllBetweenStream(
+    lowerLimit: number | null,
+    upperLimit: number | null,
+    step: number | null = 1
+  ): AsyncIterable<SignedBeaconBlock> {
     const safeLowerLimit = lowerLimit || Buffer.alloc(0);
     const safeUpperLimit = upperLimit || Number.MAX_SAFE_INTEGER;
-    const data = await this.db.search({
+    const data = this.db.searchStream({
       gt: encodeKey(this.bucket, safeLowerLimit),
       lt: encodeKey(this.bucket, safeUpperLimit),
     });
-    return (data || [])
-      .map((datum) => this.type.deserialize(datum))
-      .filter(signedBlock => {
-        if (step !== null && typeof safeLowerLimit === "number") {
-          return signedBlock.message.slot >= safeLowerLimit && (signedBlock.message.slot - safeLowerLimit) % step === 0;
-        } else {
-          return true;
+    const isEligible = (signedBlock: SignedBeaconBlock): boolean => {
+      if (step !== null && typeof safeLowerLimit === "number") {
+        return signedBlock.message.slot >= safeLowerLimit && (signedBlock.message.slot - safeLowerLimit) % step === 0;
+      } else {
+        return true;
+      }
+    };
```

Comment:
I would consider extracting this method to some other file so you can test it easily